signer/kms/aws/client.go

```diff
@@ -0,0 +1,341 @@
+func ParseReference(resourceID string) (endpoint, keyID, alias string, err error) {
+	var v []string
+	for _, re := range allREs {
+		v = re.FindStringSubmatch(resourceID)
+		if len(v) >= 3 {
+			endpoint, keyID = v[1], v[2]
+			if len(v) == 4 {
+				alias = v[3]
+			}
+			return
+		}
+	}
+	err = fmt.Errorf("invalid awskms format %q", resourceID)
+	return
+}
+
+type awsClient struct {
+	client   *akms.Client
+	endpoint string
+	keyID    string
+	alias    string
+	keyCache *ttlcache.Cache[string, cmk]
+}
+
+func newAWSClient(ctx context.Context, ksp *kms.KMSSignerProvider) (*awsClient, error) {
+	if err := ValidReference(ksp.Reference); err != nil {
+		return nil, err
+	}
+	a := &awsClient{}
+	var err error
+	a.endpoint, a.keyID, a.alias, err = ParseReference(ksp.Reference)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := a.setupClient(ctx, ksp); err != nil {
+		return nil, err
+	}
+
+	a.keyCache = ttlcache.New[string, cmk](
+		ttlcache.WithDisableTouchOnHit[string, cmk](),
+	)
+
+	return a, nil
+}
+
+func (a *awsClient) setupClient(ctx context.Context, ksp *kms.KMSSignerProvider) (err error) {
+	opts := []func(*config.LoadOptions) error{}
+	if a.endpoint != "" {
+		opts = append(opts, config.WithEndpointResolverWithOptions(
+			aws.EndpointResolverWithOptionsFunc(func(service, region string, options ...interface{}) (aws.Endpoint, error) {
+				return aws.Endpoint{
+					URL: "https://" + a.endpoint,
+				}, nil
+			}),
+		))
+	}
+	if os.Getenv("AWS_TLS_INSECURE_SKIP_VERIFY") == "1" {
+		opts = append(opts, config.WithHTTPClient(&http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{InsecureSkipVerify: true}, // nolint: gosec
+			},
+		}))
+	}
+
+	cfg, err := config.LoadDefaultConfig(ctx, opts...)
+	if err != nil {
+		return fmt.Errorf("loading AWS config: %w", err)
+	}
+
+	a.client = akms.NewFromConfig(cfg)
+	return
+}
+
+func (a *awsClient) fetchCMK(ctx context.Context) (*cmk, error) {
+	var err error
+	cmk := &cmk{}
+	cmk.PublicKey, err = a.fetchPublicKey(ctx)
+	if err != nil {
+		return nil, err
+	}
+	cmk.KeyMetadata, err = a.fetchKeyMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return cmk, nil
+}
+
+func (a *awsClient) getHashFunc(ctx context.Context) (crypto.Hash, error) {
+	cmk, err := a.getCMK(ctx)
+	if err != nil {
+		return 0, err
+	}
+	return cmk.HashFunc(), nil
+}
+
+func (a *awsClient) getCMK(ctx context.Context) (*cmk, error) {
+	var lerr error
+	loader := ttlcache.LoaderFunc[string, cmk](
+		func(c *ttlcache.Cache[string, cmk], key string) *ttlcache.Item[string, cmk] {
+			var k *cmk
+			k, lerr = a.fetchCMK(ctx)
+			if lerr == nil {
+				return c.Set(cacheKey, *k, time.Second*300)
+			}
+			return nil
+		},
+	)
+
+	item := a.keyCache.Get(cacheKey, ttlcache.WithLoader[string, cmk](loader))
+	if lerr == nil {
+		cmk := item.Value()
+		return &cmk, nil
+	}
+	return nil, lerr
+}
+
+// At the moment this function lies unused, but it is here for future if necessary
```

Comment:
food for thought: my yagni practices tell me maybe this shouldn't be here if it's unused

Reply:
yeah agreed.

Reply:
ah - this comment is now incorrect. It is used for offline verification 😄